spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/annotation/TaskListenerExecutor.java

```diff
@@ -0,0 +1,246 @@
+					if (ScopedObject.class.isAssignableFrom(type)) {
+						try {
+							type = AutoProxyUtils.determineTargetClass(this.context.getBeanFactory(),
+									ScopedProxyUtils.getTargetBeanName(beanName));
+						}
+						catch (Throwable ex) {
+							// An invalid scoped proxy arrangement - let's ignore it.
+							if (logger.isDebugEnabled()) {
+								logger.debug("Could not resolve target bean for scoped proxy '" + beanName + "'", ex);
+							}
+						}
+					}
+					try {
+						processBean(beanName, type);
+					}
+					catch (Throwable ex) {
+						throw new BeanInitializationException("Failed to process @BeforeTask " +
+								"annotation on bean with name '" + beanName + "'", ex);
+					}
+				}
+			}
+		}
+
+	}
+
+	private void processBean(String beanName, final Class<?> type){
+		if (!this.nonAnnotatedClasses.contains(type)) {
+			Map<Method, BeforeTask> beforeTaskMethods = getBeforeTaskMethods(type);
+			Map<Method, AfterTask> afterTaskMethods = getAfterTaskMethods(type);
+			Map<Method, FailedTask> failedTaskMethods = getFailedTaskMethods(type);
+
+			if (beforeTaskMethods.isEmpty() && afterTaskMethods.isEmpty()) {
+				this.nonAnnotatedClasses.add(type);
+				return;
+			}
+			if(!beforeTaskMethods.isEmpty()) {
+				this.beforeTaskMethods.putAll(beforeTaskMethods);
+				for(Method beforeTaskMethod : beforeTaskMethods.keySet()) {
+					this.beforeTaskInstances.put(beforeTaskMethod, context.getBean(beanName));
+				}
+			}
+			if(!afterTaskMethods.isEmpty()){
+				this.afterTaskMethods.putAll(afterTaskMethods);
+				for(Method afterTaskMethod : afterTaskMethods.keySet()) {
+					this.afterTaskInstances.put(afterTaskMethod, context.getBean(beanName));
+				}
+			}
+			if(!failedTaskMethods.isEmpty()){
+				this.failedTaskMethods.putAll(failedTaskMethods);
+				for(Method failedTaskMethod : failedTaskMethods.keySet()) {
+					this.failedTaskInstances.put(failedTaskMethod, context.getBean(beanName));
+				}
+			}
+		}
+	}
+
+
+	private Map<Method, BeforeTask> getBeforeTaskMethods(final Class<?> type){
+		return MethodIntrospector.selectMethods(type,
+				new MethodIntrospector.MetadataLookup<BeforeTask>() {
+					@Override
+					public BeforeTask inspect(Method method) {
+						return AnnotationUtils.findAnnotation(method, BeforeTask.class);
+					}
+				});
+	}
+
+	private Map<Method, AfterTask> getAfterTaskMethods(final Class<?> type){
+		return MethodIntrospector.selectMethods(type,
+				new MethodIntrospector.MetadataLookup<AfterTask>() {
+					@Override
+					public AfterTask inspect(Method method) {
+						return AnnotationUtils.findAnnotation(method, AfterTask.class);
+					}
+				});
+	}
+
+	private Map<Method, FailedTask> getFailedTaskMethods(final Class<?> type){
+		return MethodIntrospector.selectMethods(type,
+				new MethodIntrospector.MetadataLookup<FailedTask>() {
+					@Override
+					public FailedTask inspect(Method method) {
+						return AnnotationUtils.findAnnotation(method, FailedTask.class);
+					}
+				});
```

Comment:
If you do need these three methods (see previous comment about the `Map<Method, ANNOTATION>`, can't you do this once with generics?  Something like (I haven't tried this so feel free to call BS):
```
	public static class MethodGetter<T extends Annotation> {
		public Map<Method, T> getMethods(final Class<?> type, final Class<T> annotationClass){
			return MethodIntrospector.selectMethods(type,
					new MethodIntrospector.MetadataLookup<T>() {
						@Override
						public T inspect(Method method) {
							return AnnotationUtils.findAnnotation(method, annotationClass);
						}
					});
		}
	}
...

Map<Method, BeforeTask> beforeTask = new MethodGetter<BeforeTask>().getMethods(type, BeforeTask.class);
Map<Method, AfterTask> afterTask = new MethodGetter<AfterTask>().getMethods(type, AfterTask.class);
Map<Method, FailedTask> failedTask = new MethodGetter<FailedTask>().getMethods(type, FailedTask.class);
```